src/durmod/kaldi-durmod.h

```diff
@@ -0,0 +1,246 @@
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+#ifndef DURMOD_KALDI_DURMOD_H_
+#define DURMOD_KALDI_DURMOD_H_
+
+#include <iostream>
+#include <vector>
+#include <utility>
+#include <string>
+
+#include "base/kaldi-common.h"
+#include "nnet3/nnet-example.h"
+#include "nnet3/nnet-nnet.h"
+#include "nnet3/nnet-example-utils.h"
+#include "nnet3/nnet-optimize.h"
+#include "util/common-utils.h"
+#include "util/stl-utils.h"
+#include "fstext/deterministic-fst.h"
+
+namespace kaldi {
+  using nnet3::Nnet;
+  using nnet3::NnetExample;
+  using nnet3::NnetIo;
+  using nnet3::CachingOptimizingCompiler;
+  using nnet3::ComputationRequest;
+  using nnet3::NnetComputation;
+  using nnet3::NnetComputeOptions;
+  using nnet3::NnetComputer;
+
+struct PhoneDurationModelOptions {
+  int32 left_context, right_context;
+  int32 max_duration;
+
+  void Register(OptionsItf *po);
+
+  PhoneDurationModelOptions():
+    left_context(4),
+    right_context(2),
+    max_duration(60) {}
+};
+
+class PhoneDurationModel {
+  friend class PhoneDurationFeatureMaker;
+
+ public:
+  void Read(std::istream &is, bool binary);
+  void Write(std::ostream &os, bool binary) const;
+
+  PhoneDurationModel() {}
+  PhoneDurationModel(const PhoneDurationModelOptions &opts,
+                     const std::vector<std::vector<int32> > &roots,
+                     const std::vector<std::vector<int32> > &questions):
+    left_context_(opts.left_context),
+    right_context_(opts.right_context),
+    roots_(roots),
+    questions_(questions),
+    max_duration_(opts.max_duration) {}
+
+  inline int32 MaxDuration() const { return max_duration_; }
+  inline int32 RightContext() const { return right_context_; }
+  inline int32 LeftContext() const { return left_context_; }
+  inline int32 FullContextSize() const {
+    return left_context_ + right_context_ + 1;
+  }
+  // TODO(hhadian): this class is passive. make it struct?
+ private:
+  int32 left_context_, right_context_;
+  std::vector<std::vector<int32> > roots_;
+  std::vector<std::vector<int32> > questions_;
+  int32 max_duration_;
+};
+
+class PhoneDurationFeatureMaker {
+ public:
+  explicit PhoneDurationFeatureMaker(const PhoneDurationModel &model);
+
+  void InitFeatureMaker(const PhoneDurationModel &model);
+
+  /// This method extracts features for a phone. The inputs are a phone_context
+  /// and an index into that context which indicates the centeral phone.
+  /// The features consist of phone-IDs (which are represented in a 1-of-n
+  /// encoding, suitable for neural networks), phone durations, and binary
+  /// features determined from a set of questions (i.e. extra_questions.int)
+  void MakeFeatureVector(
+                    const std::vector<std::pair<int32, int32> > &phone_context,
```

Comment:
>OK- so it's a sequence of (phone, duration) pairs.

Yes. Github did not show part of my comment which was inside '<', '>' signs.

>Perhaps if you called it phone_dur_context it would be clearer.

Of course; I will change it.